modules/generator/processor/spanmetrics/spanmetrics.go

```diff
@@ -0,0 +1,128 @@
+package spanmetrics
+
+import (
+	"context"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+
+	gen "github.com/grafana/tempo/modules/generator/processor"
+	processor_util "github.com/grafana/tempo/modules/generator/processor/util"
+	"github.com/grafana/tempo/pkg/tempopb"
+	v1_trace "github.com/grafana/tempo/pkg/tempopb/trace/v1"
+	tempo_util "github.com/grafana/tempo/pkg/util"
+)
+
+type processor struct {
+	cfg Config
+
+	spanMetricsCallsTotal      *prometheus.CounterVec
+	spanMetricsDurationSeconds *prometheus.HistogramVec
+
+	// for testing
+	now func() time.Time
+}
+
+func New(cfg Config, tenant string) gen.Processor {
+	return &processor{
+		cfg: cfg,
+		now: time.Now,
+	}
+}
+
+func (p *processor) Name() string { return Name }
+
+func (p *processor) RegisterMetrics(reg prometheus.Registerer) error {
+	labelNames := []string{"service", "span_name", "span_kind", "span_status"}
+	if len(p.cfg.Dimensions) > 0 {
+		labelNames = append(labelNames, p.cfg.Dimensions...)
```

Comment:
I definitely think we need some limits here to keep a check on the number of active series, if we're going to allow custom dimensions. Either here or in `appender.go` where we are appending labels/samples. Maybe for a future PR though?

Reply:
Agree we need to limit the amount of series we create because this will impact the memory used by the generators. I wouldn't bother limiting it here (for now) since `dimensions` is controlled by the operator of Tempo, so it's easy to lower this when the cluster is getting in trouble.
At this point I'm more concerned about the series created by labels like `span_name` because these are dependent on the cardinality of the traces we are ingesting and we can't control that.

Limiting the active series in the appender makes a lot of sense, both to protect Tempo and to protect the downstream TSDB. I'll add it as a TODO to this PR and will make a tracking issue with all remaining tasks once this PR is merged.